File: sdk/AuthSys-Device-Example/perl/AuthSys-Device.pl
package AuthSysDevice;

use strict;
use warnings;
use LWP::UserAgent;
use JSON;
use Digest::MD5 qw(md5_hex);
use Sys::Hostname;

sub new {
    my ($class, $group_secret, $base_url) = @_;
    $base_url ||= 'https://authsys-main-production.up.railway.app/device';
    $base_url =~ s/\/+$//;
    my $self = {
        group_secret   => $group_secret,
        base_url     => $base_url,
        last_error   => '',
        last_response => '',
        ua           => LWP::UserAgent->new(timeout => 15),
    };
    $self->{ua}->default_header('Content-Type' => 'application/json');
    return bless $self, $class;
}

sub _get_hwid {
    my $raw = hostname();
    if ( $^O eq 'MSWin32' ) {
        my @mac = `getmac /nh /fo csv 2>nul`;
        for (@mac) { chomp; $raw .= $_ }
    } else {
        my @mac = `ifconfig 2>/dev/null | grep -i hwaddr`;
        for (@mac) { chomp; $raw .= $_ }
    }
    return uc(md5_hex($raw));
}

sub _request {
    my ($self, $endpoint, $payload) = @_;
    my $url = $self->{base_url} . '/' . $endpoint;
    my $json = encode_json($payload);
    my $resp = $self->{ua}->post($url, Content => $json);
    $self->{last_response} = $resp->decoded_content || '';
    return $resp;
}

sub check {
    my $self = shift;
    $self->{last_error} = '';
    my $payload = {
        group_secret => $self->{group_secret},
        hwid       => _get_hwid(),
    };
    my $resp = $self->_request('check', $payload);
    if ($resp->is_success) {
        my $data = decode_json($self->{last_response});
        if ($data->{active}) { return 1 }
        $self->{last_error} = $data->{message} || 'Device deactivated by admin';
        return 0;
    }
    $self->{last_error} = $resp->status_line;
    return 0;
}

sub register {
    my ($self, $device_name) = @_;
    $self->{last_error} = '';
    my $payload = {
        group_secret => $self->{group_secret},
        hwid       => _get_hwid(),
    };
    $payload->{device_name} = $device_name if $device_name;
    my $resp = $self->_request('register', $payload);
    if ($resp->is_success) {
        my $data = decode_json($self->{last_response});
        return $data->{active} ? 1 : 0;
    }
    $self->{last_error} = $resp->status_line;
    return 0;
}

1;

File: sdk/AuthSys-Device-Example/go/AuthSys-Device.go
package authsysdevice

import (
	"bytes"
	"crypto/md5"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

type Device struct {
	AppSecret   string
	BaseURL     string
	LastError   string
	LastResponse string
	client      *http.Client
}

func NewDevice(appSecret, baseURL string) *Device {
	if baseURL == "" {
		baseURL = "https://authsys-main-production.up.railway.app/device"
	}
	return &Device{
		AppSecret: appSecret,
		BaseURL:   strings.TrimRight(baseURL, "/"),
		client:    &http.Client{Timeout: 15 * time.Second},
	}
}

func getHWID() string {
	raw := ""
	hostname, err := os.Hostname()
	if err == nil {
		raw += hostname
	}
	interfaces, err := net.Interfaces()
	if err == nil {
		for _, iface := range interfaces {
			if len(iface.HardwareAddr) > 0 {
				raw += iface.HardwareAddr.String()
			}
		}
	}
	hash := md5.Sum([]byte(raw))
	return fmt.Sprintf("%X", hash)
}

func (d *Device) request(endpoint string, payload map[string]string) (map[string]interface{}, error) {
	body, _ := json.Marshal(payload)
	req, err := http.NewRequest("POST", d.BaseURL+"/"+endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := d.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	d.LastResponse = string(respBody)
	var result map[string]interface{}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func (d *Device) Check() bool {
	d.LastError = ""
	payload := map[string]string{
		"group_secret": d.AppSecret,
		"hwid":       getHWID(),
	}
	result, err := d.request("check", payload)
	if err != nil {
		d.LastError = err.Error()
		return false
	}
	if active, ok := result["active"].(bool); ok && active {
		return true
	}
	if msg, ok := result["message"].(string); ok {
		d.LastError = msg
	} else {
		d.LastError = "Device deactivated by admin"
	}
	return false
}

func (d *Device) Register(deviceName string) bool {
	d.LastError = ""
	payload := map[string]string{
		"group_secret": d.AppSecret,
		"hwid":       getHWID(),
	}
	if deviceName != "" {
		payload["device_name"] = deviceName
	}
	result, err := d.request("register", payload)
	if err != nil {
		d.LastError = err.Error()
		return false
	}
	active, _ := result["active"].(bool)
	return active
}
